src/lib.rs

```diff
@@ -284,9 +291,142 @@ cfg_if! {
+    imp::getrandom_inner(dest)?;
+    // SAFETY: `dest` has been fully initialized by `imp::getrandom_inner`
+    // since it returned `Ok`.
+    Ok(unsafe { slice_assume_init_mut(dest) })
+}
+
+/// A marker indicating it is safe to use `MaybeUninit::as_bytes_mut()` to
+/// initialize a valid value of the type. Feature: `rust-1-55`.
+///
+/// This trait is implemented for byte arrays (e.g. `[u8; T]`, and levels of nested byte arrays
+///
+/// Note: `TransmuteFromArbitraryBytes` is intentionally not implemented on
+/// `u8` just because it is counterproductive to generate random values one
+/// byte at a time.
+///
+/// TODO: Should this be a sealed trait.
+#[cfg(feature = "rust-1-55")]
+pub unsafe trait TransmuteFromArbitraryBytes {}
+
+#[cfg(feature = "rust-1-55")]
+unsafe impl<const N: usize> TransmuteFromArbitraryBytes for [u8; N] {}
+
+#[cfg(feature = "rust-1-55")]
+unsafe impl<T: TransmuteFromArbitraryBytes, const N: usize> TransmuteFromArbitraryBytes for [T; N] {}
+
+/// Returns a value of type `T` where all bytes of the value were
+/// initialized using `getrandom_uninit`. Feature: `rust-1-55`.
+///
+/// This works on byte arrays, without any type annotations:
+/// ```
+/// # use getrandom::getrandom_array;
+/// fn tls_hello_random() -> Result<[u8; 32], getrandom::Error> {
+///     getrandom_array()
+/// }
+/// ```
+///
+/// This also works on nested byte arrays. This nested array support can be
+/// used to safely and efficiently (in optimized builds) construct random
+/// arrays of other types, though this requires type annotations to help the
+/// compiler figure out the nested array size:
+/// ```
+/// # use getrandom::{getrandom_array, Error};
+/// # fn u32_array_example() -> Result<(), Error> {
+/// let random_u32s /* [u32; 4] */ =
+///     getrandom_array::<[_; 4]>()?.map(u32::from_ne_bytes);
+/// # Ok(())
+/// # }
+/// ```
+///
+/// Multiple levels of array nesting can be used to construct more complicated
+/// types:
+/// ```
+/// # // TODO: Use `std::simd::Simd` when the `portable_simd` feature is
+/// # // available; until then, here is a minimal polyfill for it that
+/// # // allows the examle to work in stable Rust.
+/// # struct Simd<T, const N: usize>([T; N]);
+/// # impl<T, const N: usize> From<[T; N]> for Simd<T, N> {
+/// #     fn from(value: [T; N]) -> Self {
+/// #        Self(value)
+/// #     }
+/// # }
+/// # use getrandom::{getrandom_array, Error};
+/// # fn simd_array_example() -> Result<(), Error> {
+/// let random_vectors /* [Simd<u32, 4>; 16] */ =
+///     getrandom_array::<[[_; 4]; 16]>()?.map(|bytes| {
+///         Simd::from(bytes.map(u32::from_ne_bytes))
+///     });
+/// # Ok(())
+/// # }
+/// ```
+///
+/// The number of nested arrays is limited only by whatever limits the compiler
+/// imposes:
+/// ```
+/// # use getrandom::getrandom_array;
+/// let lots_of_layers_of_nesting =
+///     getrandom_array::<[[[[[[[[[[[[[[[[u8; 1]; 1]; 1]; 1]; 1]; 1]; 1]; 1]; 1]; 1]; 1]; 1]; 1]; 1]; 1]; 1]>();
+/// ```
```

Comment:
I think you went a bit overboard with nesting, 3-4 levels should be enough to demonstrate the functionality.